protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java

```diff
@@ -119,7 +121,15 @@ public T parse(InputStream stream) {
         }
         CodedInputStream cis = null;
         try {
-          if (stream instanceof KnownLength) {
+          if (stream instanceof NioBufferBackedStream) {
+            int size = stream.available();
+            if (size > 0) {
+              NioBufferBackedStream nioBufferBackedStream = (NioBufferBackedStream) stream;
+              cis = CodedInputStream.newInstance(nioBufferBackedStream.nioBuffer());
```

Comment:
Although in `ServerStreamListenerImpl` we can assume the `InputStream` is always closed at the end, this is not a part of the `Marshaller#parse` API. If we only read some of the stream and them pass the stream in again a second time to read more data, we would start over from the beginning because `nioBuffer` returns a new one every time. To be clean here we probably should probably maintain the read offset ourselves, so that whether the user reads from the returned ByteBuffer or from reading the InputStream directly, the read offset is consistent.

I also synced with @ejona86 and learned there's a rope based API soon coming to the protobuf CodedInputStream API. I think this is the javadoc from that project:
https://github.com/google/protobuf/blob/master/java/core/src/main/java/com/google/protobuf/RopeByteString.java

The rope is essentially a list of byte arrays, which enables us to generalize your optimization to `CompositeReadableBuffer`s with more than 1 buffer, which would imply the return value of `NioBufferBackedStream#nioBuffer` will be a list of buffers.